lib/manager/index.ts

```diff
@@ -21,96 +24,36 @@ import {
   LANGUAGE_RUBY,
   LANGUAGE_RUST,
 } from '../constants/languages';
-import {
-  MANAGER_ANSIBLE,
-  MANAGER_BAZEL,
-  MANAGER_BUILDKITE,
-  MANAGER_BUNDLER,
-  MANAGER_CARGO,
-  MANAGER_CDNURL,
-  MANAGER_CIRCLE_CI,
-  MANAGER_COMPOSER,
-  MANAGER_DEPS_EDN,
-  MANAGER_DOCKER_COMPOSE,
-  MANAGER_DOCKERFILE,
-  MANAGER_DRONE_CI,
-  MANAGER_GIT_SUBMODULES,
-  MANAGER_GITHUB_ACTIONS,
-  MANAGER_GITLAB_CI,
-  MANAGER_GITLAB_CI_INCLUDE,
-  MANAGER_GO_MOD,
-  MANAGER_GRADLE,
-  MANAGER_GRADLE_WRAPPER,
-  MANAGER_HELM_REQUIREMENTS,
-  MANAGER_HELMFILE,
-  MANAGER_HOMEBREW,
-  MANAGER_KUBERNETES,
-  MANAGER_LEININGEN,
-  MANAGER_MAVEN,
-  MANAGER_METEOR,
-  MANAGER_MIX,
-  MANAGER_NPM,
-  MANAGER_NUGET,
-  MANAGER_NVM,
-  MANAGER_PIP_REQUIREMENTS,
-  MANAGER_PIP_SETUP,
-  MANAGER_PIPENV,
-  MANAGER_POETRY,
-  MANAGER_PUB,
-  MANAGER_RUBY_VERSION,
-  MANAGER_SBT,
-  MANAGER_SWIFT,
-  MANAGER_TERRAFORM,
-  MANAGER_TRAVIS,
-} from '../constants/managers';
-
-const managerList = [
-  MANAGER_ANSIBLE,
-  MANAGER_BAZEL,
-  MANAGER_BUILDKITE,
-  MANAGER_BUNDLER,
-  MANAGER_CARGO,
-  MANAGER_CDNURL,
-  MANAGER_CIRCLE_CI,
-  MANAGER_COMPOSER,
-  MANAGER_DEPS_EDN,
-  MANAGER_DOCKER_COMPOSE,
-  MANAGER_DOCKERFILE,
-  MANAGER_DRONE_CI,
-  MANAGER_GIT_SUBMODULES,
-  MANAGER_GITHUB_ACTIONS,
-  MANAGER_GITLAB_CI,
-  MANAGER_GITLAB_CI_INCLUDE,
-  MANAGER_GO_MOD,
-  MANAGER_GRADLE,
-  MANAGER_GRADLE_WRAPPER,
-  MANAGER_HELM_REQUIREMENTS,
-  MANAGER_HELMFILE,
-  MANAGER_HOMEBREW,
-  MANAGER_KUBERNETES,
-  MANAGER_LEININGEN,
-  MANAGER_MAVEN,
-  MANAGER_METEOR,
-  MANAGER_MIX,
-  MANAGER_NPM,
-  MANAGER_NUGET,
-  MANAGER_NVM,
-  MANAGER_PIP_REQUIREMENTS,
-  MANAGER_PIP_SETUP,
-  MANAGER_PIPENV,
-  MANAGER_POETRY,
-  MANAGER_PUB,
-  MANAGER_SBT,
-  MANAGER_SWIFT,
-  MANAGER_TERRAFORM,
-  MANAGER_TRAVIS,
-  MANAGER_RUBY_VERSION,
-];
 
+const managerList = [];
 const managers: Record<string, ManagerApi> = {};
-for (const manager of managerList) {
-  managers[manager] = require(`./${manager}`); // eslint-disable-line
+
+function isValidManagerModule(module: unknown): boolean {
+  // TODO: check interface and fail-fast?
+  return !!module;
+}
+
+function loadManagers(): void {
+  const managerDirs = fs
+    .readdirSync(__dirname, { withFileTypes: true })
+    .filter(dirent => dirent.isDirectory())
+    .map(dirent => dirent.name)
+    .sort();
+  for (const manager of managerDirs) {
+    let module = null;
+    try {
+      module = require(`./${manager}`); // eslint-disable-line
```

Comment:
@rarkins Shouldn't we defer load until we need that manager? That would speed up selfhosted startup time. Also most userser only need some managers

Reply:
I prefer eager loading here. I'm not convinced that the loading time would really make a difference, e.g. the total files read in from `node_modules` would dwarf all the manager files combined.

Also, I want managers to soon export their own default config (rather than put it in `lib/config/definitions.ts`) so we'd need to read them in immediately to get that anyway.

Reply:
ok